# Linear Regression: Quality & Commercial Score

Trains a Ridge regression model for each target using a `ColumnTransformer` that scales
only the continuous features. Saves the fitted pipelines and results to `results/`.

## About Linear Regression

Linear regression is the simplest prediction method: it draws a straight line through the data that gets as close as possible to every point. Each feature (like complexity, playtime, or whether a game has a particular mechanic) gets a weight — a number that says how much that feature pushes the predicted score up or down. The final prediction is just the sum of all features multiplied by their weights. We use a version called **Ridge regression**, which prevents any single feature from getting an unreasonably large weight — important when you have 400 features and some of them are correlated with each other.

![Linear Regression](../assets/diagram_linear_regression.png)

## Constants

In [32]:
# ── Constants ──────────────────────────────────────────────────────────────
import os

# How much data goes to training vs. testing.
# Default is 80/20: the model learns from 80% of games (~17,500) and we test
# on the remaining 20% (~4,400). We also run every notebook at 50/50, 70/30,
# and 90/10 to check that results stay consistent (see results/*.md).
# Configurable via environment variable so the run-notebooks script can sweep
# through all splits automatically.
TRAIN_RATIO = float(os.environ.get("LUDOMETRICS_TRAIN_RATIO", "0.80"))
TEST_RATIO = round(1.0 - TRAIN_RATIO, 10)

# Fixed seed so every run produces identical results.
# Without this, the random train/test split would change each time,
# making it impossible to compare models fairly.
RANDOM_STATE = 42

# Derived split label used in output filenames, e.g. "80_20"
SPLIT_LABEL = f"{int(TRAIN_RATIO * 100)}_{int(TEST_RATIO * 100)}"
SPLIT_DISPLAY = f"{int(TRAIN_RATIO * 100)}/{int(TEST_RATIO * 100)}"

## Prepare data

In [33]:
import sys
import time

sys.path.insert(0, "..")

import joblib
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error, r2_score

from utils.train_utils import update_results_table

df = pd.read_csv("../data/games_processed.csv")
print(f"Loaded {len(df):,} rows × {len(df.columns)} columns")

CONTINUOUS_COLS = [
    "GameWeight",
    "MinPlayers",
    "MaxPlayers",
    "ComAgeRec",
    "MfgAgeRec",
    "MfgPlaytime",
    "ComMinPlaytime",
    "ComMaxPlaytime",
]

DROP_COLS = ["BGGId", "quality_score", "commercial_score"]
FEATURE_COLS = [c for c in df.columns if c not in DROP_COLS]
TARGETS = ["quality_score", "commercial_score"]

X = df[FEATURE_COLS]
print(f"Features: {len(FEATURE_COLS)} columns ({len(CONTINUOUS_COLS)} continuous, {len(FEATURE_COLS) - len(CONTINUOUS_COLS)} binary)")

X_train, X_test = train_test_split(X, test_size=TEST_RATIO, random_state=RANDOM_STATE)
print(f"Train: {len(X_train):,}  |  Test: {len(X_test):,}")

Loaded 21,925 rows × 403 columns
Features: 400 columns (8 continuous, 392 binary)
Train: 17,540  |  Test: 4,385


## Build pipeline: StandardScaler for continuous features

Linear regression is sensitive to feature scale: a coefficient's magnitude depends on the units of its input, so features measured in minutes (0–600) would dominate features measured in player count (1–20) purely because of their range, not their importance.

`games_processed.csv` already handles imputation, capping, and joins (see `00_preprocessing.ipynb`), but intentionally leaves scaling to each training notebook because tree-based models don't need it.

Here we wrap a `StandardScaler` (zero-mean, unit-variance) into the sklearn `Pipeline` so that:
- The 8 continuous features are scaled before reaching the Ridge model.
- The 392 binary columns pass through unchanged (they're already 0/1).
- Scaling is fitted on the training set only and applied consistently at prediction time, avoiding data leakage.

In [34]:
from sklearn.compose import ColumnTransformer
from sklearn.linear_model import RidgeCV
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import StandardScaler


def build_linear_pipeline(continuous_cols: list[str], all_feature_cols: list[str]) -> Pipeline:
    """StandardScaler on continuous cols, passthrough on binary cols."""
    preprocessor = ColumnTransformer(
        transformers=[
            ("scale", StandardScaler(), continuous_cols),
        ],
        remainder="passthrough",
        verbose_feature_names_out=False,
    ).set_output(transform="pandas")

    return Pipeline(
        [
            ("preprocessor", preprocessor),
            ("model", RidgeCV()),
        ]
    )

## Train one model per target

For each target score (quality and commercial), we hand the model all 17,500 training games with their 400 features and the known score. The model figures out a weight for every feature — essentially learning "games with worker placement tend to score 2 points higher" or "longer playtime adds about 0.5 points." We feed our feature columns and targets into `build_linear_pipeline`, our helper function from above that scales the continuous features and wraps a Ridge model into a single pipeline.

The final prediction is just those 400 weights multiplied by the game's features, all added together (400 terms). A weight of +3.8 on a binary feature means having that feature adds 3.8 points to the predicted score; a weight of −2.0 means it subtracts 2 points.

In [35]:
results = {}

for target in TARGETS:
    y_train = df.loc[X_train.index, target]

    pipeline = build_linear_pipeline(CONTINUOUS_COLS, FEATURE_COLS)
    t0 = time.monotonic()
    pipeline.fit(X_train, y_train)
    elapsed = time.monotonic() - t0

    results[target] = {"pipeline": pipeline, "elapsed": elapsed}
    print(f"{target:20s}  trained in {elapsed:.1f}s")

# Show what the model learned: the features with the largest weights.
# Each weight tells you how many points that feature adds (+) or subtracts (−)
# from the predicted score.
for target in TARGETS:
    pipeline = results[target]["pipeline"]
    model = pipeline.named_steps["model"]
    feature_names = pipeline.named_steps["preprocessor"].get_feature_names_out()
    top_idx = np.argsort(np.abs(model.coef_))[::-1][:5]
    print(f"\n{target} — top 5 highest-weight features:")
    for i in top_idx:
        print(f"  {feature_names[i]:30s}  {model.coef_[i]:+.1f} points")

quality_score         trained in 0.3s
commercial_score      trained in 0.4s

quality_score — top 5 highest-weight features:
  TableauBuilding                 +3.8 points
  Delayed Purchase                +3.4 points
  Advantage Token                 +3.1 points
  Cat:Strategy                    +3.1 points
  Legacy Game                     +3.0 points

commercial_score — top 5 highest-weight features:
  Cat:Party                       +14.0 points
  Cat:Family                      +12.8 points
  Cat:Strategy                    +10.7 points
  Cat:Thematic                    +9.4 points
  Bingo                           +8.6 points


## Evaluate

Now we check if the model actually learned something real. We pick a well-known game from the test set and ask the model to predict its scores, comparing the prediction against the actual value shows us how close the model gets on a single game. Then we evaluate across all 4,400 test games using **RMSE** (the average prediction error in points out of 100, lower is better) and **R²** (what fraction of score variation the model explains: 1.0 is perfect, 0.0 means it learned nothing).

In [36]:
# Load game names for readable output
game_names = pd.read_csv("../dataset/games.csv", usecols=["BGGId", "Name"]).set_index("BGGId")["Name"]

# Pick the highest-rated game in the test set as our example
example_idx = df.loc[X_test.index, "quality_score"].idxmax()
example_row = X_test.loc[[example_idx]]
example_id = df.loc[example_idx, "BGGId"]
example_name = game_names.get(example_id, f"BGGId {example_id}")

print(f"Example prediction ({example_name}):")
for target in TARGETS:
    pipeline = results[target]["pipeline"]
    pred = pipeline.predict(example_row)[0]
    actual = df.loc[example_idx, target]
    print(f"  Predicted {target}: {pred:.1f}")
    print(f"  Actual    {target}: {actual:.1f}")
print()

# Evaluate on all test games
for target in TARGETS:
    # Commonly referred to as y_test — the actual target values for the test set
    y_actual = df.loc[X_test.index, target]
    pipeline = results[target]["pipeline"]
    # Commonly referred to as y_pred — the model's predicted values
    y_predicted = pipeline.predict(X_test)
    rmse = mean_squared_error(y_actual, y_predicted) ** 0.5
    r2 = r2_score(y_actual, y_predicted)
    results[target]["rmse"] = rmse
    results[target]["r2"] = r2
    results[target]["y_predicted"] = y_predicted
    results[target]["y_actual"] = y_actual
    print(f"{target:20s}  RMSE={rmse:.4f}  R²={r2:.4f}")

Example prediction (Gloomhaven):
  Predicted quality_score: 81.5
  Actual    quality_score: 85.1
  Predicted commercial_score: 109.4
  Actual    commercial_score: 94.3

quality_score         RMSE=2.7841  R²=0.5124
commercial_score      RMSE=10.3130  R²=0.4964


## Save results

In [37]:
MODEL_DIR = Path("../results/models/linear_regression")
MODEL_DIR.mkdir(parents=True, exist_ok=True)

for target in TARGETS:
    out_path = MODEL_DIR / f"{target}_{SPLIT_LABEL}.pkl"
    joblib.dump(results[target]["pipeline"], out_path)
    print(f"Saved {out_path}")

    predictions = pd.DataFrame({
        "BGGId": df.loc[X_test.index, "BGGId"].values,
        "actual": results[target]["y_actual"].values,
        "predicted": results[target]["y_predicted"],
    })
    pred_path = MODEL_DIR / f"{target}_{SPLIT_LABEL}_predictions.csv"
    predictions.to_csv(pred_path, index=False)
    print(f"Saved {pred_path}")

RESULTS_DIR = Path("../results")

for target in TARGETS:
    elapsed = results[target]["elapsed"]
    minutes, seconds = divmod(elapsed, 60)
    time_label = (
        f"{int(minutes)}m {seconds:.1f}s" if minutes >= 1 else f"{seconds:.1f}s"
    )
    update_results_table(
        RESULTS_DIR / f"{target}.md",
        algorithm="Linear Regression (RidgeCV)",
        split=SPLIT_DISPLAY,
        train_size=len(X_train),
        test_size=len(X_test),
        training_time=time_label,
        rmse=results[target]["rmse"],
        r2=results[target]["r2"],
    )
    print(f"Updated results/{target}.md")

Saved ../results/models/linear_regression/quality_score_80_20.pkl
Saved ../results/models/linear_regression/quality_score_80_20_predictions.csv
Saved ../results/models/linear_regression/commercial_score_80_20.pkl
Saved ../results/models/linear_regression/commercial_score_80_20_predictions.csv
Updated results/quality_score.md
Updated results/commercial_score.md
